smslibrary/src/main/java/com/eis/smslibrary/SMSReceivedBroadcastReceiver.java

```diff
@@ -30,30 +33,36 @@ public void onReceive(Context context, Intent intent) {
         if (extras == null) {
             return;
         }
-
-        Object[] smsExtras = (Object[]) extras.get("pdus");
+        Object[] pdus = (Object[]) extras.get("pdus");
+        if (pdus == null) return;
         String format = (String) extras.get("format");
+        Map<String, String> peerMessageMap = new HashMap<>(pdus.length);
 
-        if(smsExtras == null) //could be null
-            return;
-
-        for (Object smsData : smsExtras) {
-            SmsMessage receivedSMS = createMessageFromPdu(smsData, format);
-            String smsContent = receivedSMS.getMessageBody();
-            String phoneNumber = receivedSMS.getOriginatingAddress();
-
-            if (phoneNumber == null) //could be null
-                return;
-
-            SMSMessage parsedMessage = SMSMessageHandler.getInstance().parseMessage(phoneNumber, smsContent);
-            if (parsedMessage != null) {
-                callApplicationService(context, parsedMessage);
+        /* There can be multiple SMSes from multiple senders. We'll concatenate multiple SMSes from
+         * the same sender in one message.
+         */
+        for (Object pdu : pdus) {
+            SmsMessage message = createMessageFromPdu(pdu, format);
+            String phoneNumber = message.getOriginatingAddress();
+            if (phoneNumber == null) phoneNumber = "";
+            if (!peerMessageMap.containsKey(phoneNumber)) {
+                peerMessageMap.put(phoneNumber, message.getMessageBody());
+            } else {
+                String previousParts = peerMessageMap.get(phoneNumber);
+                String updatedMessage = previousParts + message.getMessageBody();
```

Comment:
are you sure that parts are received in order?

Reply:
I remember reading somewhere in the documentation that intents received in BroadcastReceivers with intent filter `android.provider.Telephony.SMS_RECEIVED` will contain a bundle of PDUs containing all parts of multi-part text messages received, I'll try to find the link and post it here

Reply:
ok that's fine

Reply:
[Here](https://developer.android.com/reference/android/provider/Telephony.Sms.Intents#SMS_RECEIVED_ACTION) it is:
> The intent will have the following extra values:
>- "pdus" - An Object[] of byte[]s containing the PDUs that make up the message.

Reply:
yes

Reply:
You're right, since Android doesn't provide any methods to get CSMS reference numbers, total number of parts and part number from PDUs, and the Intent contains all PDUs that make up the message, I assumed that the Android system would order them itself before giving us the bundle of PDUs. I'll check the source code of Android's default messaging app to see how they're handled there.

Reply:
I read the [source code](https://android.googlesource.com/platform/packages/apps/Messaging/) of AOSP's default messaging app and this is how the `SmsMessage[]` array is built:
```java
            /**
             * Read the PDUs out of an {@link #SMS_RECEIVED_ACTION} or a
             * {@link #DATA_SMS_RECEIVED_ACTION} intent.
             *
             * @param intent the intent to read from
             * @return an array of SmsMessages for the PDUs
             */
            public static SmsMessage[] getMessagesFromIntent(Intent intent) {
                Object[] messages;
                try {
                    messages = (Object[]) intent.getSerializableExtra("pdus");
                }
                catch (ClassCastException e) {
                    Rlog.e(TAG, "getMessagesFromIntent: " + e);
                    return null;
                }

                if (messages == null) {
                    Rlog.e(TAG, "pdus does not exist in the intent");
                    return null;
                }

                String format = intent.getStringExtra("format");
                int subId = intent.getIntExtra(PhoneConstants.SUBSCRIPTION_KEY,
                        SubscriptionManager.getDefaultSmsSubscriptionId());

                Rlog.v(TAG, " getMessagesFromIntent sub_id : " + subId);

                int pduCount = messages.length;
                SmsMessage[] msgs = new SmsMessage[pduCount];

                for (int i = 0; i < pduCount; i++) {
                    byte[] pdu = (byte[]) messages[i];
                    msgs[i] = SmsMessage.createFromPdu(pdu, format);
                    if (msgs[i] != null) msgs[i].setSubId(subId);
                }
                return msgs;
            }
```

An this is how the message body is built from `SmsMessage[]`:
```java
    // Parse the message body from message PDUs
    private static String buildMessageBodyFromPdus(final SmsMessage[] msgs) {
        if (msgs.length == 1) {
            // There is only one part, so grab the body directly.
            return replaceFormFeeds(msgs[0].getDisplayMessageBody());
        } else {
            // Build up the body from the parts.
            final StringBuilder body = new StringBuilder();
            for (final SmsMessage msg : msgs) {
                try {
                    // getDisplayMessageBody() can NPE if mWrappedMessage inside is null.
                    body.append(msg.getDisplayMessageBody());
                } catch (final NullPointerException e) {
                    // Nothing to do
                }
            }
            return replaceFormFeeds(body.toString());
        }
    }
```
Code of method `replaceFormFeeds()`:

```java
    // Some providers send formfeeds in their messages. Convert those formfeeds to newlines.
    private static String replaceFormFeeds(final String s) {
        return s == null ? "" : s.replace('\f', '\n');
    }
```

I don't have time to check if this works the same way as in my `BroadcastReceiver`, I'm writing this comment just to avoid having to search all of this again tomorrow

Reply:
il looks like they already come in order

Reply:
Yes, it also looks like it's not possible to get multiple messages from different senders in the same `Intent`. I don't know where I got that idea from. I'll remove the `HashMap` I was using to separate messages from different senders.

Reply:
yes, you can mark this as resolved